### PR TITLE
Add ssoIssuer to fix Response issuer checking

### LINKED ISSUE #884: SAML issuer not the same for request and response
AuthnRequest issuer refers to the service provider.
Response issuer refers to the identity provider.

This means the issuer check at https://github.com/coreos/dex/blob/master/connector/saml/saml.go#L288 isn't correct.

`p.issuer` is the SP, and `resp.Issuer.Issuer` is the IDP.

Spec: http://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-tech-overview-2.0.html

```xml
<samlp:AuthnRequest
xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
ID="identifier_1"
Version="2.0"
IssueInstant="2004-12-05T09:21:59Z"
AssertionConsumerServiceIndex="1">
<saml:Issuer>https://sp.example.com/SAML2</saml:Issuer>
<samlp:NameIDPolicy
AllowCreate="true"
Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"/>
</samlp:AuthnRequest>
```

```xml
<samlp:Response
xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
ID="identifier_2"
InResponseTo="identifier_1"
Version="2.0"
IssueInstant="2004-12-05T09:22:05Z"
Destination="https://sp.example.com/SAML2/SSO/POST">
<saml:Issuer>https://idp.example.org/SAML2</saml:Issuer>
<samlp:Status>
<samlp:StatusCode
Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
</samlp:Status>
<saml:Assertion
xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
ID="identifier_3"
Version="2.0"
IssueInstant="2004-12-05T09:22:05Z">
<saml:Issuer>https://idp.example.org/SAML2</saml:Issuer>
<!-- a POSTed assertion MUST be signed -->
<ds:Signature
xmlns:ds="http://www.w3.org/2000/09/xmldsig#">...</ds:Signature>
<saml:Subject>
<saml:NameID
Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient">
3f7b3dcf-1674-4ecd-92c8-1544f346baf8
</saml:NameID>
<saml:SubjectConfirmation
Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
<saml:SubjectConfirmationData
InResponseTo="identifier_1"
Recipient="https://sp.example.com/SAML2/SSO/POST"
NotOnOrAfter="2004-12-05T09:27:05Z"/>
</saml:SubjectConfirmation>
</saml:Subject>
<saml:Conditions
NotBefore="2004-12-05T09:17:05Z"
NotOnOrAfter="2004-12-05T09:27:05Z">
<saml:AudienceRestriction>
<saml:Audience>https://sp.example.com/SAML2</saml:Audience>
</saml:AudienceRestriction>
</saml:Conditions>
<saml:AuthnStatement
AuthnInstant="2004-12-05T09:22:00Z"
SessionIndex="identifier_3">
<saml:AuthnContext>
<saml:AuthnContextClassRef>
urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport
</saml:AuthnContextClassRef>
</saml:AuthnContext>
</saml:AuthnStatement>
</saml:Assertion>
</samlp:Response>
```

### PR DESCRIPTION
**Pull request by Calpicow**
Fixes #884

### CODE REVIEW

**Comment by Calpicow** [comment]
Just realized we already have AudienceRestriction checking in https://github.com/coreos/dex/blob/master/connector/saml/saml.go#L462. Should we remove the checking altogether?

File: connector/saml/saml.go [hunk 1]
```diff
@@ -81,8 +81,10 @@ type Config struct {
 	//
 	// https://www.oasis-open.org/committees/download.php/35391/sstc-saml-metadata-errata-2.0-wd-04-diff.pdf
 
-	Issuer string `json:"issuer"`
-	SSOURL string `json:"ssoURL"`
+	// Issuer refers to the SP issuer
+	Issuer    string `json:"issuer"`
+	SSOIssuer string `json:"ssoIssuer"`
```
Comment (ericchiang) [inline connector/saml/saml.go]: I'm a little confused by this fix. What's SSOIssuer? How is it different than issuer? What part of the spec is enforcing this?

Comment (Calpicow) [inline connector/saml/saml.go]: Refer to the issue I filed. In a nutshell, there appears to be some confusion between the SP issuer and IDP issuer.

Comment (ericchiang) [inline connector/saml/saml.go]: Okay thanks. Give me a few to read up on your linked article.

Comment (Calpicow) [inline connector/saml/saml.go]: Any updates? We don't really need to add this field, and just remove the Response issuer checking if you like.

Comment (ericchiang) [inline connector/saml/saml.go]: Sorry, we've been dealing with a security issue in this connector. Please see https://groups.google.com/forum/#!topic/coreos-user/s_BWh4a3E1Y

File: connector/saml/saml_test.go [hunk 2]
```diff
@@ -62,17 +62,17 @@ func runVerify(t *testing.T, ca string, resp string, shouldSucceed bool) {
...
+	if ssoIssuer == "" {
+		ssoIssuer = defaultSSOIssuer
 	}
 	if redirectURI == "" {
 		redirectURI = defaultRedirectURI
 	}
 	now, _ := time.Parse(time.RFC3339, "2017-01-24T20:48:41Z")
 	timeFunc := func() time.Time { return now }
 	return &provider{
-		issuer:       issuer,
```
Comment (ericchiang) [inline connector/saml/saml_test.go]: Sorry, trying to re-review this. Issuer is no longer being used here then, correct? I was under the impression that `issuer` was supposed to be the value that you've changed to `ssoIssuer` to anyway. E.g. that `issuer` should just be `http://www.okta.com/exk91cb99lKkKSYoy0h7`.

Maybe this an issue with the documentation?

Comment (ericchiang) [inline connector/saml/saml_test.go]: Ah, nvm I see what you're saying.

**Comment by ericchiang** [comment]
Couple comments:

So we might want to rename issuer in this case, since it has different meanings for OpenID Connect. Issuer means the remote provider, not the dex server. Maybe `SSOIssuer` and `EntityIssuer`?

Also is this always going to be the same or different then redirectURI?

**Comment by Calpicow** [comment]
It is possible for issuer it to be different than redirectURI. I've avoided renaming issuer since it would be a breaking change, but I'm open to that.

**Comment by ericchiang** [comment]
> I've avoided renaming issuer since it would be a breaking change, but I'm open to that.

We've marked it experimental explicitly so we can do these kind of breaking changes :)

**Comment by ericchiang** [comment]
(to give a timeline, we'd like to do one final push over the next couple days to get it to stable)

**Comment by Calpicow** [comment]
Done

**Pull request merged**